lnrpc/routerpc/router_server.go

```diff
@@ -0,0 +1,265 @@
+	// We make sure that we register it with the main gRPC server to ensure
+	// all our methods are routed properly.
+	RegisterRouterServer(grpcServer, s)
+
+	log.Debugf("Router RPC server successfully register with root gRPC " +
+		"server")
+
+	return nil
+}
+
+// SendPayment attempts to route a payment described by the passed
+// PaymentRequest to the final destination. If we are unable to route the
+// payment, or cannot find a route that satisfies the constraints in the
+// PaymentRequest, then an error will be returned. Otherwise, the payment
+// pre-image, along with the final route will be returned.
+func (s *Server) SendPayment(ctx context.Context,
+	req *PaymentRequest) (*PaymentResponse, error) {
+
+	switch {
+	// If the payment request isn't populated, then we won't be able to
+	// even attempt a payment.
+	case req.PayReq == "":
+		return nil, fmt.Errorf("a valid payment request MUST be specified")
+	}
+
+	// Now that we know the payment request is present, we'll attempt to
+	// decode it in order to parse out all the parameters for the route.
+	payReq, err := zpay32.Decode(req.PayReq, s.cfg.ActiveNetParams)
+	if err != nil {
+		return nil, err
+	}
+
+	// Atm, this service does not support invoices that don't have their
+	// value fully specified.
+	if payReq.MilliSat == nil {
+		return nil, fmt.Errorf("zero value invoices are not supported")
+	}
+
+	// Now that all the information we need has been parsed, we'll map this
+	// proto request into a proper request that our backing router can
+	// understand.
+	finalDelta := uint16(payReq.MinFinalCLTVExpiry())
+	payment := routing.LightningPayment{
+		Target:            payReq.Destination,
+		Amount:            *payReq.MilliSat,
+		FeeLimit:          lnwire.MilliSatoshi(req.FeeLimit),
+		PaymentHash:       *payReq.PaymentHash,
+		FinalCLTVDelta:    &finalDelta,
+		PayAttemptTimeout: time.Second * time.Duration(req.TimeoutSeconds),
+		RouteHints:        payReq.RouteHints,
+	}
+
+	preImage, _, err := s.cfg.Router.SendPayment(&payment)
+	if err != nil {
+		return nil, err
+	}
+
+	return &PaymentResponse{
+		PayHash:  (*payReq.PaymentHash)[:],
+		PreImage: preImage[:],
+	}, nil
+}
+
+// EstimateRouteFee allows callers to obtain a lower bound w.r.t how much it
+// may cost to send an HTLC to the target end destination.
+func (s *Server) EstimateRouteFee(ctx context.Context,
+	req *RouteFeeRequest) (*RouteFeeResponse, error) {
+
+	// First we'll parse out the raw public key into a value that we can
+	// utilize.
+	destNode, err := btcec.ParsePubKey(req.Dest, btcec.S256())
+	if err != nil {
+		return nil, err
+	}
+
+	// Next, we'll convert the amount in satoshis to mSAT, which are the
+	// native unit of LN.
+	amtMsat := lnwire.NewMSatFromSatoshis(btcutil.Amount(req.AmtSat))
+
+	// Finally, we'll query for a route to the destination that can carry
+	// that target amount, we'll only request a single route.
+	routes, err := s.cfg.Router.FindRoutes(
```

Comment:
Imo we should not increase the dependency on the k-shortest algorithm as we might remove it in the future. Because it is a single route, we can use `findPath` here directly?

Reply:
It would indeed require adding a `FindRoute` function that generates the bandwidth hints before calling `findPath`. But I don't see much more is needed other than that.

Reply:
Indeed, this PR is focused on just adding the sub-server. Adding a singular `FindRoute` (along with adding tests, etc) will need to be when/if we decide to deprecate `FindRoutes` all together. I imagine that eventually we'll not even attempt a regular query, and instead may want more of a fee range to be returned to give the caller an idea w.r.t both the min and max.